### ISSUE TITLE
failed to prepare subPath for volumeMount "minio-server-config" of container "minio"

### ISSUE THREAD

**Issue by unicman** [minio/minio#6076]
Minio pods fail to run giving following error - failed to prepare subPath for volumeMount "minio-server-config" of container "minio".

## Expected Behavior
When minio is tried to be installed through spinnaker helm chart, all the pods should eventually reach 'Running' state including minio.

## Current Behavior
All other pods go to 'Running' state but minio pod remains in CreateContainerConfigError state forever. Tried with images - minio/minio:RELEASE.2018-04-27T23-33-52Z and minio/minio:RELEASE.2018-03-16T22-52-12Z.

kubectl describe pod keeps showing following error -

  Type     Reason  Age                   From               Message
  ----     ------  ----                  ----               -------
  Normal   Pulled  45m (x18516 over 2d)  kubelet, slc12obc  Container image "minio/minio:RELEASE.2018-03-16T22-52-12Z" already present on machine
  Warning  Failed  24s (x18721 over 2d)  kubelet, slc12obc  Error: failed to prepare subPath for volumeMount "minio-server-config" of container "minio"

## Possible Solution
If this is already known issue... need a workaround / reference to image / configuration that will work with kubernetes 1.9.1. version. I can't change Kubernetes version just to make minio work.

## Steps to Reproduce (for bugs)
Install docker, kubernetes and helm.

`[root@slc12obc tmp]# docker version
Client:
 Version:      17.06.2-ol
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   d02b7ab
 Built:        Fri Oct  6 00:02:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ol
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   d02b7ab
 Built:        Fri Oct  6 00:03:48 2017
 OS/Arch:      linux/amd64
 Experimental: false

[root@slc12obc tmp]# kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1+2.1.5.el7", GitCommit:"ea4c95b07dd46d5563e4f6851ede13157bf04266", GitTreeState:"archive", BuildDate:"2018-03-23T02:54:15Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1+2.1.5.el7", GitCommit:"ea4c95b07dd46d5563e4f6851ede13157bf04266", GitTreeState:"archive", BuildDate:"2018-03-23T02:54:15Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

[root@slc12obc tmp]# helm version
Client: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}`

1. Create 3 persistent volumes required for spinnaker.
2. Run command helm install  -f /tmp/spinnaker-values.yml --name my-spinnaker stable/spinnaker --namespace spinnaker --timeout 1500 --wait
3. This command will time out and later kubectl get all --namespace spinnaker shows following.

`
NAME                                        DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deploy/my-spinnaker-jenkins                 1         1         1            1           2d
deploy/my-spinnaker-minio                   1         2         1            0           2d
deploy/my-spinnaker-redis                   1         1         1            1           2d
deploy/my-spinnaker-spinnaker-clouddriver   1         1         1            0           2d
deploy/my-spinnaker-spinnaker-deck          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-echo          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-front50       1         1         1            0           2d
deploy/my-spinnaker-spinnaker-gate          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-igor          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-orca          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-rosco         1         1         1            1           2d

NAME                                               DESIRED   CURRENT   READY     AGE
rs/my-spinnaker-jenkins-75c675686b                 1         1         1         2d
rs/my-spinnaker-minio-5b7466b656                   1         1         0         2d
rs/my-spinnaker-minio-66c55784dc                   1         1         0         2d
rs/my-spinnaker-redis-59d485f4cc                   1         1         1         2d
rs/my-spinnaker-spinnaker-clouddriver-55456cb645   1         1         0         2d
rs/my-spinnaker-spinnaker-deck-6b9c7f7775          1         1         1         2d
rs/my-spinnaker-spinnaker-echo-66cf6d6855          1         1         1         2d
rs/my-spinnaker-spinnaker-front50-57f5f88857       1         1         0         2d
rs/my-spinnaker-spinnaker-gate-55fc5c56ff          1         1         1         2d
rs/my-spinnaker-spinnaker-igor-8646895db8          1         1         1         2d
rs/my-spinnaker-spinnaker-orca-8d6bd475d           1         1         1         2d
rs/my-spinnaker-spinnaker-rosco-cdd47bdc8          1         1         1         2d

NAME                                        DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
deploy/my-spinnaker-jenkins                 1         1         1            1           2d
deploy/my-spinnaker-minio                   1         2         1            0           2d
deploy/my-spinnaker-redis                   1         1         1            1           2d
deploy/my-spinnaker-spinnaker-clouddriver   1         1         1            0           2d
deploy/my-spinnaker-spinnaker-deck          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-echo          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-front50       1         1         1            0           2d
deploy/my-spinnaker-spinnaker-gate          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-igor          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-orca          1         1         1            1           2d
deploy/my-spinnaker-spinnaker-rosco         1         1         1            1           2d

NAME                                               DESIRED   CURRENT   READY     AGE
rs/my-spinnaker-jenkins-75c675686b                 1         1         1         2d
rs/my-spinnaker-minio-5b7466b656                   1         1         0         2d
rs/my-spinnaker-minio-66c55784dc                   1         1         0         2d
rs/my-spinnaker-redis-59d485f4cc                   1         1         1         2d
rs/my-spinnaker-spinnaker-clouddriver-55456cb645   1         1         0         2d
rs/my-spinnaker-spinnaker-deck-6b9c7f7775          1         1         1         2d
rs/my-spinnaker-spinnaker-echo-66cf6d6855          1         1         1         2d
rs/my-spinnaker-spinnaker-front50-57f5f88857       1         1         0         2d
rs/my-spinnaker-spinnaker-gate-55fc5c56ff          1         1         1         2d
rs/my-spinnaker-spinnaker-igor-8646895db8          1         1         1         2d
rs/my-spinnaker-spinnaker-orca-8d6bd475d           1         1         1         2d
rs/my-spinnaker-spinnaker-rosco-cdd47bdc8          1         1         1         2d

NAME                                                     READY     STATUS                       RESTARTS   AGE
po/my-spinnaker-jenkins-75c675686b-5xsg4                 1/1       Running                      0          2d
po/my-spinnaker-minio-5b7466b656-bbx5m                   0/1       CreateContainerConfigError   0          2d
po/my-spinnaker-minio-66c55784dc-v2zmb                   0/1       CreateContainerConfigError   0          2d
po/my-spinnaker-redis-59d485f4cc-6wd68                   1/1       Running                      0          2d
po/my-spinnaker-spinnaker-clouddriver-55456cb645-htbc9   0/1       Running                      0          2d
po/my-spinnaker-spinnaker-deck-6b9c7f7775-zmtn7          1/1       Running                      0          2d
po/my-spinnaker-spinnaker-echo-66cf6d6855-hf4s4          1/1       Running                      0          2d
po/my-spinnaker-spinnaker-front50-57f5f88857-kxvd2       0/1       Running                      779        2d
po/my-spinnaker-spinnaker-gate-55fc5c56ff-r7rhx          1/1       Running                      2          2d
po/my-spinnaker-spinnaker-igor-8646895db8-bcttz          1/1       Running                      0          2d
po/my-spinnaker-spinnaker-orca-8d6bd475d-bvpwk           1/1       Running                      0          2d
po/my-spinnaker-spinnaker-rosco-cdd47bdc8-xs2lf          1/1       Running                      0          2d

NAME                                     TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                       AGE
svc/my-spinnaker-jenkins                 NodePort    110.1.16.196    <none>        8080:31531/TCP                2d
svc/my-spinnaker-jenkins-agent           ClusterIP   110.1.74.237    <none>        50000/TCP                     2d
svc/my-spinnaker-minio                   ClusterIP   None             <none>        9000/TCP                      2d
svc/my-spinnaker-redis                   ClusterIP   110.1.251.169   <none>        6379/TCP                      2d
svc/my-spinnaker-spinnaker-clouddriver   ClusterIP   110.1.106.27    <none>        7002/TCP                      2d
svc/my-spinnaker-spinnaker-deck          NodePort    110.1.96.251     <none>        9000:31434/TCP                2d
svc/my-spinnaker-spinnaker-echo          ClusterIP   110.1.80.53     <none>        8089/TCP                      2d
svc/my-spinnaker-spinnaker-front50       ClusterIP   110.1.160.75    <none>        8080/TCP                      2d
svc/my-spinnaker-spinnaker-gate          ClusterIP   110.1.155.244   <none>        8084/TCP                      2d
svc/my-spinnaker-spinnaker-igor          ClusterIP   110.1.153.121   <none>        8088/TCP,8080/TCP,50000/TCP   2d
svc/my-spinnaker-spinnaker-orca          ClusterIP   110.1.77.159     <none>        8083/TCP                      2d
svc/my-spinnaker-spinnaker-rosco         ClusterIP   110.1.243.238   <none>        8087/TCP                      2d
`

## Context
Just because of error in minio, spinnaker doesn't work.

## Your Environment
* Version used (`minio version`): minio/minio:RELEASE.2018-04-27T23-33-52Z and minio/minio:RELEASE.2018-03-16T22-52-12Z.
* Environment name and version (e.g. nginx 1.9.1): Docker 17.06.02-ol, Kubernetes 1.9.1, Helm 2.8.0.
* Server type and version: Oracle Linux 7.4 update 4
* Operating System and version (`uname -a`): Linux host_name 4.1.12-112.14.10.el7uek.bug27766149.x86_64 #2 SMP Mon Apr 2 09:25:05 PDT 2018 x86_64 x86_64 x86_64 GNU/Linux
* Link to your project: N.A.

**Comment by nitisht** [comment]
what is the helm minio chart version you are using @unicman ? Can you try using [Minio Chart version](https://github.com/kubernetes/charts/tree/master/stable/minio) `1.3.4`

**Comment by unicman** [comment]
I've not modified any of the charts so not sure exactly.

Helm requirements.yaml of spinnaker at https://github.com/kubernetes/charts/blob/master/stable/spinnaker/requirements.yaml says following,

`- name: minio
  version: 1.1.1
  repository: https://kubernetes-charts.storage.googleapis.com/
  condition: minio.enabled`

Does that help?

**Comment by nitisht** [comment]
yes, but this chart is pretty old and uses configMap (which I think is the reason for the failure you see). Can you update the version to `1.3.4` and give it a go?

**Comment by unicman** [comment]
It's part of spinnaker chart definition... not sure how I will be able to change that version. Are you suggesting that I should just make code change and commit in this Git repo?

**Comment by unicman** [comment]
Do you know any easier way without making code change in spinnaker chart?

**Comment by nitisht** [comment]
Will it be possible to make the changes and test locally (no need to push to GitHub)? I can then tell you the steps..

**Comment by unicman** [comment]
Helm downloads latest charts from github and runs them when 'helm install...' is called. I don't think there is a way to make local changes to it and have helm use it. At least I don't know of such facility in helm.

**Comment by nitisht** [comment]
> I don't think there is a way to make local changes to it and have helm use it. At least I don't know of such facility in helm.

Helm has an [inbuilt server](https://github.com/kubernetes/helm/blob/master/docs/helm/helm_serve.md) that can serve charts from anywhere. To test locally you can 

- Change the version in Spinnaker chart. 
- Use `helm package <chart-name>` to package the chart. ([ref](https://github.com/kubernetes/helm/blob/master/docs/helm/helm_package.md))
- Start helm serve from the directory where you have helm chart package.
- Add a repo `local` to your helm cli using [helm repo add](https://github.com/kubernetes/helm/blob/master/docs/helm/helm_repo_add.md) command.
- Do a `helm install local/minio` (instead of usual `helm install stable/minio).

**Comment by unicman** [comment]
Haven't done it before. Let me see if that's possible because 'helm' tool itself does git clone internally and not sure where it keeps the files.

**Comment by nitisht** [comment]
Sure @unicman , meanwhile I have sent a PR to Spinnaker Helm chart so it uses the current Minio helm chart. https://github.com/kubernetes/charts/pull/6368

**Comment by unicman** [comment]
Oh that's great! BTW i didn't know you already had that access... so created separate ticket for them - https://github.com/kubernetes/charts/issues/6351.

**Comment by nitisht** [comment]
@unicman I am going to close this for now, as there is nothing to be done from Minio server code perspective. 

Once the PR kubernetes/charts#6368 is merged, this should work, if not, feel free to reopen this.

**Comment by sboagibm** [comment]
@nitisht It's still not clear to me how to install minio on Kubernetes v1.10.0 (helm version v2.7.2), minio-0.5.2.  Sorry if the answer is somewhere in the discussion above.

**Comment by nitisht** [comment]
you can refer this doc @sboagibm https://docs.minio.io/docs/deploy-minio-on-kubernetes.html

**Comment by CloudWarGit** [comment]
@nitisht I quickly go trough the source code of minio server initconfig part , it call os.Rename to config.json file, so that will be failed when config.json file is mounted by configmap. I think this is not a grace way for user which want manage their config file by configmap.

**Comment by nitisht** [comment]
@CloudWarGit As configmaps are mounted readonly, we mount it on `/tmp/config.json` and then move the config file to Minio config directory. 
https://github.com/helm/charts/blob/master/stable/minio/templates/deployment.yaml#L53

Additionally Minio is moving to a flat config file design which can be simply updated using `mc` sub commands. So, in coming releases of Helm Chart we'll completely remove configmaps.
